Add tracks event for "View product" and dismiss action after update

nathanss opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Add tracks event for "View product" and dismiss action after update

Closes #35494 .

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Edit a product
2. Click "Update" to submit the changes
3. You should see a banner like this: 
<img width="743" alt="image" src="https://user-images.githubusercontent.com/13437655/201729495-39723901-3219-472e-a1cb-871ee2d4a1a2.png">
4. Enable tracks logging by typing `localStorage.setItem( 'debug', '*' );` in the console
5. Click "View product"
6. Check that the tracks event was fired accordingly.
7. Make the banner appear again by updating again, but this time click the X icon to dismiss it.
8. Check that the tracks event was fired accordingly

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

nathanss committed:
Record an event 'product_view_product_click' when the 'View Product' link is clicked after creating or updating a product

nathanss committed:
Add product_view_product_dismiss tracks events (WIP, not currently working)

nathanss committed:
Add changelog

nathanss committed:
Fix whitespace in addEventListener and remove console.logs

mattsherman commented:
> Also, could my small PHP change be the reason unit tests are breaking?

They appear unrelated. You can always try re-running the railing test. If it still fails, you can re-base against the latest `trunk` in hopes that the failing test was fixed and you just got unlucky when you originally branched from `trunk`.

mattsherman commented on plugins/woocommerce/includes/admin/class-wc-admin-post-types.php:
```diff
@@ -125,13 +125,13 @@ public function post_updated_messages( $messages ) {
 			0  => '', // Unused. Messages start at index 1.
 			/* translators: %s: Product view URL. */
-			1  => sprintf( __( 'Product updated. <a href="%s">View Product</a>', 'woocommerce' ), esc_url( get_permalink( $post->ID ) ) ),
+			1  => sprintf( __( 'Product updated. <a id="view-product-link" href="%s">View Product</a>', 'woocommerce' ), esc_url( get_permalink( $post->ID ) ) ),
```
I'm not sure if we have any guidelines around `id` naming, but this seems a bit generic... what is an extension already added an element with this `id`?

Maybe it should be a bit more specific, something like `woocommerce-product-updated-message-view-product-link`? (double check our CSS naming guideslines... I imagine that isn't quite right)

joshuatf replied:
One more thing I'll add here is that it might be beneficial to move the class name outside of the localized string.  This has 2 distinct advantages:

1. We prevent breaking of translations between versions if we ever need to update or add to the class.
2. It reduces human error from translators in case they misspell a class name.

`<a href="%1$s" class="%2$s">`

Another option is to move the entire link tag out of the localized string, which works well and seems favorable, but for some reason we rarely seem to do it:

https://github.com/woocommerce/woocommerce/blob/34bf2df5e680f84cbb552c654205bfe0ffd8cdf9/plugins/woocommerce/src/Internal/Admin/Notes/AddFirstProduct.php#L70

mattsherman replied:
That's a nice approach!

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -324,4 +324,21 @@ export const initProductScreenTracks = () => {
+		} );
+
+	const dismissProductUpdatedButtonSelector =
+		'.notice-success.is-dismissible > button';
```
So, not supported well enough by all browsers to actually do, but..

```css
p:has(#view-product-link) ~ .notice-dismiss
```

Would be a bit more robust, since we are in control of the `view-product-link` id (see comment about that name below).

Without `:has()` though, we would have to query for the `#view-product-link` and then use the DOM to get the parent, and then the sibling button.

Not sure that is any more robust. Any solution depends on the DOM for the notice not changing, and we are not in control of that.

mattsherman commented:
> I am not super excited about the dismiss button selector. It seems very fragile. But I couldn't find a better alternative, since it doesn't have any specific id and it's WordPress standard code. Would appreciate some suggestion regarding this!

Yeah, I agree that it is fragile. I left a comment with another thought about getting the button. But, it is not widely enough supported, and traversing the DOM would also be equally fragile.

The good thing about this fragility is that if it were to break, it would only be breaking the tracking (so, our metrics). From a user standpoint, nothing would fail. We'd see a drop in the metric and wonder what was up. Not ideal, but probably good enough.

mattsherman reviewed (changes requested):
Left a few more thoughts and suggestions.

joshuatf commented on plugins/woocommerce/includes/admin/class-wc-admin-post-types.php:
```diff
@@ -125,13 +125,13 @@ public function post_updated_messages( $messages ) {
 			5  => __( 'Revision restored.', 'woocommerce' ),
 			/* translators: %s: product url */
-			6  => sprintf( __( 'Product published. <a href="%s">View Product</a>', 'woocommerce' ), esc_url( get_permalink( $post->ID ) ) ),
+			6  => sprintf( __( 'Product published. <a id="view-product-link" href="%s">View Product</a>', 'woocommerce' ), esc_url( get_permalink( $post->ID ) ) ),
```
Same comment here.

nathanss committed:
Change link tag id to be more specific and comply with CSS guidelines
Move HTML code out of translation function

mattsherman commented on plugins/woocommerce/includes/admin/class-wc-admin-post-types.php:
```diff
@@ -124,14 +124,14 @@ public function post_updated_messages( $messages ) {
 			0  => '', // Unused. Messages start at index 1.
-			/* translators: %s: Product view URL. */
-			1  => sprintf( __( 'Product updated. <a href="%s">View Product</a>', 'woocommerce' ), esc_url( get_permalink( $post->ID ) ) ),
+			/* translators: %1$s: Product link opening tag. %2$s: Product link closing tag.*/
+			1  => sprintf(__('Product updated. %1$sView Product%2$s', 'woocommerce'), '<a id="woocommerce-product-updated-message-view-product__link" href="' . esc_url(get_permalink($post->ID)) . '">', '</a>'),
```
The formatting is off here... see the PHPCS errors.

You should make sure your dev env is setup to auto-format files when saving, to ensure that code style issues are handled for you.

mattsherman commented on plugins/woocommerce/includes/admin/class-wc-admin-post-types.php:
```diff
@@ -124,14 +124,14 @@ public function post_updated_messages( $messages ) {
 			5  => __( 'Revision restored.', 'woocommerce' ),
-			/* translators: %s: product url */
-			6  => sprintf( __( 'Product published. <a href="%s">View Product</a>', 'woocommerce' ), esc_url( get_permalink( $post->ID ) ) ),
+			/* translators: %1$s: Product link opening tag. %2$s: Product link closing tag.*/
+			6  => sprintf(__('Product published. %1$sView Product%2$s', 'woocommerce'), '<a id="woocommerce-product-updated-message-view-product__link" href="' . esc_url(get_permalink($post->ID)) . '">', '</a>'),
```
Same comment as above.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -324,4 +324,23 @@ export const initProductScreenTracks = () => {
+			'#woocommerce-product-updated-message-view-product__link'
+		)
+		?.addEventListener( 'click', () => {
+			recordEvent( 'product_view_product_click', getProductData() );
```
`getProductData()` is not returning all the properties we need to, to match `wcadmin_product_update`. See acceptance criteria on #35494.

You'll want to update `getProductData()` to include the same fields as in `wcadmin_product_update`.

See https://github.com/woocommerce/woocommerce/blob/19058ed17c63f5cae9b49d0694ebff8db0dccc42/plugins/woocommerce/includes/tracks/events/class-wc-products-tracking.php#L164

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -324,4 +324,23 @@ export const initProductScreenTracks = () => {
+		document
+			.querySelector( dismissProductUpdatedButtonSelector )
+			?.addEventListener( 'click', () => {
+				recordEvent( 'product_view_product_dismiss', getProductData() );
```
Same comment as above.

mattsherman reviewed (changes requested):
Almost there! Tested, and overall works well.

Missing some product properties in the Tracks event.

And, some minor code formatting issues causing checks to fail.

nathanss committed:
Run phpcs in class-wc-admin-post-types

nathanss committed:
Add tinyMCE types to add global tinymce variable in typescript code

nathanss committed:
Add additional data in getProductData

TODO: I'm not sure if it's possible, or if it makes sense, to include stock_quantity_update in the object

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -14,7 +14,40 @@ import { waitUntilElementIsPresent } from './utils';
+		document.querySelectorAll( '.block-editor' )[ 0 ]
+	) {
+		isBlockEditor = true;
+	}
```
Couldn't you just do the following?

```
const isBlockEditor = document.querySelectorAll( '.block-editor' ).length > 0;
```

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -14,7 +14,40 @@ import { waitUntilElementIsPresent } from './utils';
+				'[name="tax_input[product_tag]"]'
+			) as HTMLInputElement
+		 ).value;
+		if ( jQuery( '#content' ).is( ':visible' ) ) {
```
I think the following would be equivalent...

```
! ( window.getComputedStyle( document.querySelector( '#content' ) ).display === 'none' )
```

Via: https://makandracards.com/makandra/1339-check-whether-an-element-is-visible-or-hidden-with-javascript

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -324,4 +416,23 @@ export const initProductScreenTracks = () => {
 			}
 		} );
+
+	document
```
You could get the initial stock value here and do something similar as was done in `track_product_updated_client_side` to log `stock_quantity_update`.

You could then either pass the initial stock value into `getProductData()`, or just store it in a const in the outer scope that contains `getProductData()` so that you could access it.

mattsherman commented on pnpm-lock.yaml:
```diff
@@ -13077,7 +13079,7 @@ packages:
       '@babel/core': 7.17.8
       postcss: 7.0.39
-      postcss-syntax: 0.36.2_postcss@7.0.39
+      postcss-syntax: 0.36.2_postcss@8.4.12
```
Why all these other changes in `pnpm-lock.yaml`? I'll admit I dont' deeply understand the lockfile, so perhaps this is no problem? I would have expected to have only seen the `@types/tinymce` change.

You should run this by somebody else on the team to make sure this is okay.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -29,7 +62,66 @@ const getProductData = () => {
+			document.querySelector( '#menu_order' ) as HTMLInputElement
+		 )?.value
+			? 'Yes'
+			: 'No',
```
Similar to how I address this in https://github.com/woocommerce/woocommerce/pull/35605, this should be something like

```
menu_order: parseInt( 
    (document.querySelector( '#menu_order' ) as HTMLInputElement)?.value ?? 0, 10 ) !== 0 ? 'Yes' : 'No',
```

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -29,7 +62,66 @@ const getProductData = () => {
+			? 'Yes'
+			: 'No',
+		weight: ( document.querySelector( '#_weight' ) as HTMLInputElement )
+			.value
```
should be `?.value`

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -29,7 +62,66 @@ const getProductData = () => {
+		description: description_value.trim() !== '' ? 'Yes' : 'No',
+		enable_reviews: (
+			document.querySelector( '#comment_status' ) as HTMLInputElement
+		 ).checked
```
Should be `?.checked`

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -29,7 +62,66 @@ const getProductData = () => {
 		manage_stock: (
 			document.querySelector( '#_manage_stock' ) as HTMLInputElement
 		 )?.value,
```
The values recorded for:

* `is_downloadable`
* `is_virtual`
* `manage_stock`

are all incorrect.

They always record as `on`, since that is the `value` of the corresponding input field.

Instead, we should be checking the `.checked` property.

We should correct this since we are right in this code now, and let @pmcpinto know that any analysis of the following Tracks events using those properties prior to this fix are invalid:

* `wcadmin_product_copy`
* `wcadmin_product_delete`
* `wcadmin_product_publish_widget_edit`
* `wcadmin_product_publish_widget_save`

pmcpinto replied:
@mattsherman thanks for the heads up. I agree that the issues should be fixed asap. I'm assuming this is affecting wcadmin_product_update and wcadmin_product_add_publish, correct?

mattsherman replied:
@pmcpinto No, I believe this only affects the Tracks events I listed.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -29,7 +62,66 @@ const getProductData = () => {
+						'#_thumbnail_id'
+					) as HTMLInputElement
+				 )?.value,
+				2
```
The radix should be `10`

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -29,7 +62,66 @@ const getProductData = () => {
+			.value
+			? 'Yes'
+			: 'No',
+		// stock_quantity_update: TODO
```
See suggestion below

mattsherman reviewed (changes requested):
A number of new properties are being logged incorrectly.

I also noted a few existing properties that are being logged incorrectly.

As well, I gave a suggestion on how `stock_quantity_update` could be handled.

nathanss committed:
Add code review suggestions, including:

* Create isElementVisible function to backfill jQuery implementation
* Fix wrong value for is_downloadable, is_virtual, manage_stock
* Fix wrong radix
* Update menu_order implementation
* Add missing ? to weight

nathanss committed:
Add stock_quantity_update implementation which is not working as intended

Since the page reloads after product update, the initialStockValue is updated as well

mattsherman commented:
> I tried implementing something similar to `track_product_updated_client_side`. But it doesn't seem to work as intended, since the page reloads.

Oh, of course. 😄

> Example:
> 
> 1. open the page. initialStockValue = 5
> 2. update stock to 7
> 3. click on "Update" and page reloads, initialStockValue is now 7
> 4. "View product" banner appears.
> 5. I click on "View product".
> 6. initialStockValue and currentStockValue will both be 7, so `stock_quantity_update` will be `No`.
> 
> @mattsherman Is that the behavior we want?

No.

If we can't obtain the previous stock value after the page reload, we should just drop the `stock_quantity_update` prop altogether.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -13,23 +13,127 @@ import { waitUntilElementIsPresent } from './utils';
 		is_downloadable: (
 			document.querySelector( '#_downloadable' ) as HTMLInputElement
-		 )?.value,
+		 )?.checked,
```
`wcadmin_product_update` records this as `No` or `Yes`. We should do the same here for consistency.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -13,23 +13,127 @@ import { waitUntilElementIsPresent } from './utils';
 		is_virtual: (
 			document.querySelector( '#_virtual' ) as HTMLInputElement
-		 )?.value,
+		 )?.checked,
```
`wcadmin_product_update` records this as `No` or `Yes`. We should do the same here for consistency.

mattsherman commented on plugins/woocommerce-admin/client/wp-admin-scripts/product-tracking/shared.ts:
```diff
@@ -13,23 +13,127 @@ import { waitUntilElementIsPresent } from './utils';
 		manage_stock: (
 			document.querySelector( '#_manage_stock' ) as HTMLInputElement
-		 )?.value,
+		 )?.checked,
```
`wcadmin_product_update` records this as `No` or `Yes`. We should do the same here for consistency.

mattsherman commented:
> > I tried implementing something similar to `track_product_updated_client_side`. But it doesn't seem to work as intended, since the page reloads.
> 
> Oh, of course. 😄
> 
> > Example:
> > 
> > 1. open the page. initialStockValue = 5
> > 2. update stock to 7
> > 3. click on "Update" and page reloads, initialStockValue is now 7
> > 4. "View product" banner appears.
> > 5. I click on "View product".
> > 6. initialStockValue and currentStockValue will both be 7, so `stock_quantity_update` will be `No`.
> > 
> > @mattsherman Is that the behavior we want?
> 
> No.
> 
> If we can't obtain the previous stock value after the page reload, we should just drop the `stock_quantity_update` prop altogether.

We could potentially pass along the previous stock value in the request to update, but that feels a bit messy and too specific to this scenario.

@pmcpinto Would it be okay if we just didn't include `stock_quantity_udpate` on these new Tracks events (when the banner is interacted with)?

mattsherman reviewed (changes requested):
@nathanss I feel like we are really close!

I left a few comments to made things consistent with existing Tracks events.

And, the comment about `stock_quantity_update`... I think you could just remove that from this PR and if we need it, it can be added in a follow up PR, since it is more involved.

nathanss committed:
Remove stock_quantity_update property and adjust is_downloadable, is_virtual, and manage_stock for consistency

mattsherman reviewed (approved):
Thanks for all the updates and tweaks, @nathanss! Nice work. I'm approving this. As long as the tests all pass after re-running, this is good to go.

If you need to rebase against the latest `trunk` in order to get things passing, let me know and I'll re-review again.

pmcpinto commented:
> @pmcpinto Would it be okay if we just didn't include stock_quantity_udpate on these new Tracks events (when the banner is interacted with)?

Yes. Thanks
